### PR TITLE
features: Add install failure tracking removal through `spack clean`

### PR DESCRIPTION
**Pull request by tldahlgren**
**Note this PR addresses a feature a user asked about when #13100 was under testing.**

The distributed build process automatically removes failure tracking information for relevant packages when ``spack install`` is executed.  This has been shown to cause problems with environment installations from ``spack.yaml`` files -- see #15415 -- that led to the need to optionally retain failure tracking.

This PR adds the ability to explicitly remove install failure tracking data for a Spack instance through the `spack clean` command using either the `-f` or `-a` options.

TODO
------
- [x] Add option to documentation

Follow-On Work
----------------
- If #15415 is merged before this PR,  `packaging_guide.rst`'s discussion of `--keep-failures` should be changed to reference the `spack clean -f` option

### CODE REVIEW

File: lib/spack/spack/database.py [hunk 1]
```diff
@@ -365,6 +366,32 @@ def _failed_spec_path(self, spec):
...
+        tty.debug('Releasing prefix failure locks')
+        for pkg_id in list(self._prefix_failures.keys()):
+            lock = self._prefix_failures.pop(pkg_id, None)
+            if lock:
+                lock.release_write()
+
+        tty.debug('Removing the prefix failure lock file')
+        try:
+            os.remove(self.prefix_fail_path)
+        except Exception:
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: I don't think it should be necessary to remove the lock file.  This potentially breaks installs in progress, and the lockfile shouldn't require cleanup -- it is just one file.  The locks are dropped when their respective processes die.

Also in general we shouldn't be catching `Exception` -- is there a reason to do it here?

File: lib/spack/spack/database.py [hunk 2]
```diff
@@ -365,6 +366,32 @@ def _failed_spec_path(self, spec):
...
+            pass
+
+        tty.debug('Ensure have an (unlocked) prefix failure lock file')
+        fs.touch(self.prefix_fail_path)
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: this can be removed.

File: lib/spack/spack/database.py [hunk 3]
```diff
@@ -365,6 +366,32 @@ def _failed_spec_path(self, spec):
...
+
+        # Remove all failure markings (aka files)
+        tty.debug('Removing prefix failure tracking files')
+        for fail_mark in os.listdir(self._failure_dir):
+            try:
+                os.remove(os.path.join(self._failure_dir, fail_mark))
+            except Exception as exc:
```
Comment (tgamblin) [inline lib/spack/spack/database.py]: Use a narrower exception class.  `OSError` at least.

Also, there should probably be a permission check here -- the removing user will need write access to the failure directory to remove the file.  Do we set it up that way?  If this requires a lot of thought we should save the permission check for a future PR.

File: lib/spack/spack/installer.py [hunk 4]
```diff
@@ -374,6 +374,13 @@ def _update_explicit_entry_in_db(pkg, rec, explicit):
             rec.explicit = True
 
 
+def clear_failures():
+    """
+    Remove all failure tracking markers for the Spack instance.
+    """
+    spack.store.db.clear_all_failures()
```
Comment (tgamblin) [inline lib/spack/spack/installer.py:380]: not opposed to this function but it does seem to indicate that the failure file management should get its own class.

Comment (tldahlgren) [reply lib/spack/spack/installer.py:380]: True.  Where does the refactor fit within the priorities?

File: lib/spack/spack/database.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/database.py]: Okay.  Great point about permissions.

File: lib/spack/spack/database.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/database.py]: I'll assume you mean all four new lines.

File: lib/spack/spack/database.py (hunk 1 above)
Comment (tldahlgren) [reply lib/spack/spack/database.py]: I was trying to be thorough wrt removing all traces of failure tracking.

Comment (tldahlgren) [reply lib/spack/spack/database.py]: Removed.

File: lib/spack/spack/database.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/database.py]: Removed.

File: lib/spack/spack/database.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/database.py]: There's a separate PR for permissions-related work.  See #16614

Comment (tldahlgren) [reply lib/spack/spack/database.py]: Changed exception to `OSError`.

**Comment by tldahlgren** [comment]
FWIW. Merged `develop` since local testing was failing for me due to assertions related to missing provider entries in `etc/spack/defaults/packages.yaml`.  Post-merge local tests then had issues with `test_make_elf_binaries_relative` (fixed in a separate PR) and `test_clean_default`.

**Comment by tldahlgren** [comment]
Note project and patch coverage diffs show two uncovered lines that were unchanged in by this PR.

**Comment by tldahlgren** [comment]
Interesting about the foreground/background test failing on MacOS (see below).  I've seen this on other PRs as well.

```
=================================== FAILURES ===================================
________ test_foreground_background_output[test_fn1-termios_on_or_off1] ________

test_fn = <function mock_shell_v_v_no_termios at 0x10b296dd0>
capfd = <_pytest.capture.CaptureFixture object at 0x11ac22bd0>
termios_on_or_off = <function no_termios at 0x10b296b90>
tmpdir = local('/private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/pytest-of-runner/pytest-0/test_foreground_background_out1')

    @pytest.mark.skipif(not which("ps"), reason="requires ps utility")
    @pytest.mark.skipif(not termios, reason="requires termios support")
    @pytest.mark.parametrize('test_fn,termios_on_or_off', [
        (mock_shell_v_v, nullcontext),
        (mock_shell_v_v_no_termios, no_termios),
    ])
    def test_foreground_background_output(
            test_fn, capfd, termios_on_or_off, tmpdir):
```

**Comment by tldahlgren** [comment]
Oops.  That was supposed to be a comment, not closing the PR.

**Comment by tldahlgren** [comment]
@becker33 The coverage "issue" is with untouched code near the changes.

**Pull request merged**